File: tests/nonzero.py
# Copyright 2021 NVIDIA Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import numpy as np

import cunumeric as num


def assert_equal(numarr, nparr):
    for resultnp, resultnum in zip(nparr, numarr):
        assert np.array_equal(resultnp, resultnum)


def test():
    assert num.count_nonzero(num.array([])) == 0
    assert num.count_nonzero(num.array([], dtype="?")) == 0
    assert_equal(num.nonzero([]), np.nonzero([]))

    assert num.count_nonzero(num.array([[], []])) == 0
    assert num.count_nonzero(num.array([[], []], dtype="?")) == 0
    assert_equal(num.nonzero([[], []]), np.nonzero([[], []]))

    assert num.count_nonzero(num.array([0])) == 0
    assert num.count_nonzero(num.array([0], dtype="?")) == 0
    assert_equal(num.nonzero(num.array([0])), ([],))

    assert num.count_nonzero(num.array([1])) == 1
    assert num.count_nonzero(num.array([1], dtype="?")) == 1
    assert_equal(num.nonzero([1]), np.nonzero([1]))

    assert num.count_nonzero(num.array(0)) == 0
    assert num.count_nonzero(num.array(0, dtype="?")) == 0
    assert_equal(num.nonzero(0), np.nonzero(0))

    assert num.count_nonzero(num.array(1)) == 1
    assert num.count_nonzero(num.array(1, dtype="?")) == 1
    assert_equal(num.nonzero(1), np.nonzero(1))

    x = num.array([1, 0, 2, -1, 0, 0, 8])
    x_np = np.array([1, 0, 2, -1, 0, 0, 8])
    assert num.count_nonzero(x) == 4
    assert_equal(num.nonzero(x), np.nonzero(x_np))

    # TBD: this will work once nonzero returns output in row-major output
    # x = num.array([[0, 1, 0], [2, 0, 3]])
    # x_np = np.array([[0, 1, 0], [2, 0, 3]])
    # assert (num.count_nonzero(x) == 3)
    # assert_equal(num.nonzero(x), np.nonzero(x_np))

    x_lg = num.eye(3)
    x_np = np.eye(3)
    assert num.count_nonzero(x_lg) == np.count_nonzero(x_np)
    assert_equal(num.nonzero(x_lg), np.nonzero(x_np))

    # TBD: this will work once nonzero returns output in row-major output
    # x = num.array([[[0, 1], [1, 1], [7, 0], [1, 0], [0, 1]], [[3, 0], [0, 3], [0, 0], [2, 2], [0, 19]]]) # noqa E501
    # x_np = np.array([[[0, 1], [1, 1], [7, 0], [1, 0], [0, 1]], [[3, 0], [0, 3], [0, 0], [2, 2], [0, 19]]]) # noqa E501
    # assert (num.count_nonzero(x) == np.count_nonzero(x_np))
    # assert_equal(num.count_nonzero(x, axis=0), np.count_nonzero(x_np, axis=0)) # noqa E501
    # assert_equal(num.count_nonzero(x, axis=1), np.count_nonzero(x_np, axis=1)) # noqa E501
    # assert_equal(num.count_nonzero(x, axis=2), np.count_nonzero(x_np, axis=2)) # noqa E501
    # assert (num.count_nonzero(x, axis=(0, 1, 2)) == np.count_nonzero(x_np, axis=(0, 1, 2))) # noqa E501
    # assert_equal(num.nonzero(x), np.nonzero(x_np))

    # x_np = np.concatenate((x_np,) * 2000, axis=1)
    # x = num.array(x_np)
    # assert (num.count_nonzero(x) == np.count_nonzero(x_np))
    # assert_equal(num.count_nonzero(x, axis=0), np.count_nonzero(x_np, axis=0)) # noqa E501
    # assert_equal(num.count_nonzero(x, axis=1), np.count_nonzero(x_np, axis=1)) # noqa E501
    # assert_equal(num.count_nonzero(x, axis=2), np.count_nonzero(x_np, axis=2)) # noqa E501
    # assert (num.count_nonzero(x, axis=(0, 1, 2)) == np.count_nonzero(x_np, axis=(0, 1, 2))) # noqa E501
    # assert_equal(num.nonzero(x), np.nonzero(x_np))

    # lg_nonzero = num.nonzero(x)
    # np_nonzero = np.nonzero(x_np)
    # assert_equal(lg_nonzero, np_nonzero)

    assert_equal(num.nonzero(0), ([],))
    assert_equal(num.nonzero(1), ([0],))

    x_np = np.random.randn(100)
    indices = np.random.choice(
        np.arange(x_np.size), replace=False, size=int(x_np.size * 0.2)
    )
    x_np[indices] = 0
    x = num.array(x_np)
    assert num.count_nonzero(x) == np.count_nonzero(x_np)
    lg_nonzero = num.nonzero(x)
    np_nonzero = np.nonzero(x_np)
    assert_equal(lg_nonzero, np_nonzero)

    # x_np = x_np.reshape(10, 10)
    # x = num.array(x_np)
    # assert (num.count_nonzero(x) == np.count_nonzero(x_np))
    # lg_nonzero = num.nonzero(x)
    # np_nonzero = np.nonzero(x_np)
    # assert_equal(lg_nonzero, np_nonzero)

    return


if __name__ == "__main__":
    test()

File: src/cunumeric/search/nonzero.cc
/* Copyright 2021-2022 NVIDIA Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "cunumeric/search/nonzero.h"
#include "cunumeric/search/nonzero_template.inl"

namespace cunumeric {

using namespace legate;

template <Type::Code CODE, int32_t DIM>
struct NonzeroImplBody<VariantKind::CPU, CODE, DIM> {
  using VAL = legate_type_of<CODE>;

  void operator()(std::vector<Array>& outputs,
                  const AccessorRO<VAL, DIM>& in,
                  const Pitches<DIM - 1>& pitches,
                  const Rect<DIM>& rect,
                  const size_t volume)
  {
    int64_t size = 0;

    for (size_t idx = 0; idx < volume; ++idx) {
      auto point = pitches.unflatten(idx, rect.lo);
      size += in[point] != VAL(0);
    }

    std::vector<Buffer<int64_t>> results;
    for (auto& output : outputs)
      results.push_back(output.create_output_buffer<int64_t, 1>(Point<1>(size), true));

    int64_t out_idx = 0;
    for (size_t idx = 0; idx < volume; ++idx) {
      auto point = pitches.unflatten(idx, rect.lo);
      if (in[point] == VAL(0)) continue;
      for (int32_t dim = 0; dim < DIM; ++dim) results[dim][out_idx] = point[dim];
      ++out_idx;
    }
    assert(size == out_idx);
  }
};

/*static*/ void NonzeroTask::cpu_variant(TaskContext& context)
{
  nonzero_template<VariantKind::CPU>(context);
}

namespace  // unnamed
{
static void __attribute__((constructor)) register_tasks(void) { NonzeroTask::register_variants(); }
}  // namespace

}  // namespace cunumeric
